File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.rb
# @param {String} s
# @return {Integer}
def my_atoi(s)
  i, sign, result = 0, 1, 0
  s = s.lstrip
  if i < s.length && ['+', '-'].include?(s[i])
    sign = -1 if s[i] == '-'
    i += 1
  end
  while i < s.length && s[i] =~ /\d/
    digit = s[i].to_i
    if result > (2**31 - 1 - digit) / 10
      return sign == 1 ? 2**31 - 1 : -2**31
    end
    result = result * 10 + digit
    i += 1
  end
  sign * result
end

File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
    let i = 0, sign = 1, result = 0;
    while (s[i] === ' ') i++;
    if (s[i] === '+' || s[i] === '-') sign = s[i++] === '-' ? -1 : 1;
    while (s[i] >= '0' && s[i] <= '9') {
        let digit = s[i++] - '0';
        if (result > (2**31 - 1 - digit) / 10)
            return sign === 1 ? 2**31 - 1 : -(2**31);
        result = result * 10 + digit;
    }
    return result * sign;
};